### PR TITLE
BWS-78: + backend and frontend

### PR DESCRIPTION
**Pull request by ania-leichenko**
Screenshot_6

### CODE REVIEW

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  useEffect,
+  useMemo,
+} from 'hooks/hooks';
+import { getNameOf } from 'helpers/helpers';
+import {
+  EAMTenantUpdateRequestDto,
+  EAMTenantFormDto,
+} from 'common/types/types';
+import { eamTenantUpdate as EAMTenantValidationSchema } from 'validation-schemas/validation-schemas';
+import { app as EAMTenantConfigurateActions } from 'store/actions';
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx]: this is part of the `eam` page. so move all actions there

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  const defaultValues = useMemo(() => {
+    const values = {
+      ...DEFAULT_PAYLOAD,
+    };
+    if (tenant) {
+      values.name = tenant.name;
+    }
+    return values;
+  }, [tenant]);
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx]:
```suggestion
  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant), [tenant]);
```
create this helper `mapTenantDefaultValues`

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  const handleFormSubmit = (payload: EAMTenantFormDto): void => {
+    if (tenant) {
+      dispatch(
+        EAMTenantConfigurateActions.updateTenant({
+          ...payload,
+          id: tenant.id,
+        }),
+      );
+    }
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx]:
```suggestion
      dispatch(
        EAMTenantConfigurateActions.updateTenant({
          ...payload,
          id: (tenant as Tenant).id,
        }),
      );
```

File: frontend/src/hooks/use-app-form/use-app-form.hook.ts [hunk 4]
```diff
@@ -1,4 +1,9 @@
-import { useForm, UseFormHandleSubmit } from 'react-hook-form';
+import {
+  useForm,
+  UseFormHandleSubmit,
+  UseFormReset,
+  FieldValues,
+} from 'react-hook-form';
```
Comment (what1s1ove) [inline frontend/src/hooks/use-app-form/use-app-form.hook.ts]: use type from our types folder

File: frontend/src/hooks/use-app-form/use-app-form.hook.ts [hunk 5]
```diff
@@ -16,6 +21,7 @@ type UseAppFormResult<T extends FormControlValues = FormControlValues> = {
   control: FormControl;
   errors: FormControlErrors;
   handleSubmit: UseFormHandleSubmit<T>;
+  reset: UseFormReset<FieldValues>;
```
Comment (what1s1ove) [inline frontend/src/hooks/use-app-form/use-app-form.hook.ts]:
```suggestion
  handleReset: UseFormReset<FieldValues>;
```

File: frontend/src/hooks/use-app-form/use-app-form.hook.ts [hunk 6]
```diff
@@ -37,6 +44,7 @@ const useAppForm = <T extends FormControlValues = FormControlValues>({
     control,
     errors,
     handleSubmit: handleSubmit as UseFormHandleSubmit<T>,
+    reset,
```
Comment (what1s1ove) [inline frontend/src/hooks/use-app-form/use-app-form.hook.ts]:
```suggestion
    handleReset: reset,
```

File: frontend/src/services/tenant-api/tenant-api.service.ts [hunk 7]
```diff
@@ -31,6 +32,19 @@ class TenantApi {
       },
     );
   }
+
+  public updateTenant(
+    params: EAMTenantUpdateRequestDto,
+  ): Promise<EAMTenantByIdResponseDto> {
+    return this.#http.load(
+      joinItems(this.#apiPrefix, ApiPath.TENANTS, '/', params.id),
```
Comment (what1s1ove) [inline frontend/src/services/tenant-api/tenant-api.service.ts]:
```suggestion
      joinItems(this.#apiPrefix, ApiPath.TENANTS, TenantsApiPath.ROOT, params.id),
```

File: frontend/src/store/app/actions.ts [hunk 8]
```diff
@@ -18,4 +19,15 @@ const getTenant = createAsyncThunk<
...
 
-export { getTenant };
+const updateTenant = createAsyncThunk<
+  EAMTenantByIdResponseDto,
+  EAMTenantUpdateRequestDto,
+  AsyncThunkConfig
+>(ActionType.UPDATE_TENANT_NAME, async (payload, { extra }) => {
+  const { tenantApi, notification } = extra;
+  const tenant = await tenantApi.updateTenant(payload);
+  notification.success('Success!', 'Tenant has been successfully created');
```
Comment (what1s1ove) [inline frontend/src/store/app/actions.ts]: we have enums for it. use them

File: frontend/src/store/app/actions.ts [hunk 9]
```diff
@@ -18,4 +19,15 @@ const getTenant = createAsyncThunk<
   return tenant;
 });
 
-export { getTenant };
+const updateTenant = createAsyncThunk<
```
Comment (what1s1ove) [inline frontend/src/store/app/actions.ts]: action should be in the `eam` page scope

File: frontend/src/store/app/common.ts [hunk 10]
```diff
@@ -1,5 +1,6 @@
 enum ActionType {
   GET_TENANT = 'app/get-tenant',
+  UPDATE_TENANT_NAME = 'eam-configurate-tenant/update-tenant-name',
```
Comment (what1s1ove) [inline frontend/src/store/app/common.ts]:
```suggestion
```

File: shared/src/validation/eam-tenant-update/eam-tenant-update.validation-schema.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  EAMTenantValidationMessage,
+  EAMTenantValidationRule,
+} from '~/common/enums/enums';
+
+const eamTenantUpdate = Joi.object({
+  [getNameOf<EAMTenantUpdateRequestDto>('name')]: Joi.string()
+    .trim()
+    .min(EAMTenantValidationRule.NAME_MIN_LENGTH)
+    .max(EAMTenantValidationRule.NAME_MAX_LENGTH)
+    // .regex(EAMTenantValidationRule.NAME_REGEX)
```
Comment (what1s1ove) [inline shared/src/validation/eam-tenant-update/eam-tenant-update.validation-schema.ts]: remove comments

**Review by what1s1ove — CHANGES_REQUESTED** [review]
image
lets add title. something like "Change Tenant Name" as title and "Tenant Name" as label for input

File: frontend/src/components/eam/components/tenant/tenant.tsx (hunk 3 above)
Comment (ania-leichenko) [reply frontend/src/components/eam/components/tenant/tenant.tsx]: where do we get this Tenant?

File: frontend/src/hooks/use-app-form/use-app-form.hook.ts (hunk 4 above)
Comment (ania-leichenko) [reply frontend/src/hooks/use-app-form/use-app-form.hook.ts]: Do you want me to transfer types?

File: frontend/src/store/app/actions.ts (hunk 8 above)
Comment (ania-leichenko) [reply frontend/src/store/app/actions.ts]: What do you mean?, I don't quite understand, tell me please

File: frontend/src/components/eam/components/tenant/tenant.tsx (hunk 3 above)
Comment (what1s1ove) [reply frontend/src/components/eam/components/tenant/tenant.tsx]: I mean a type that we store in the redux store

File: frontend/src/hooks/use-app-form/use-app-form.hook.ts (hunk 4 above)
Comment (what1s1ove) [reply frontend/src/hooks/use-app-form/use-app-form.hook.ts]: Only that we don't have
most of them we already have

File: frontend/src/store/app/actions.ts (hunk 8 above)
Comment (what1s1ove) [reply frontend/src/store/app/actions.ts]: Check an example - https://github.com/BinaryStudioAcademy/bsa-winter-2021-2022-bws/blob/development/frontend/src/store/eam-group-configurate/actions.ts#L35

Comment (ania-leichenko) [reply frontend/src/store/app/actions.ts]: thanks)

File: backend/src/data/repositories/tenant/tenant.repository.ts [hunk 12]
```diff
@@ -26,6 +26,17 @@ class Tenant {
     return Tenant.modelToEntity(tenant);
   }
 
+  async updateById(id: string, name: string): Promise<TenantEntity | null> {
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/tenant/tenant.repository.ts]: Repositories should not contain `updateById`-like methods. Here is a quick explanation:
Often when we need to update something we also decide to have some validation before updating. After some time these validations overwhelm repository methods and make them unsopportable.
If you want update data you need to create update methods on Entities and add `save(tenant: TenantEntity)` method to repository

File: backend/src/services/tenant/tenant.service.ts [hunk 13]
```diff
@@ -32,6 +32,25 @@ class Tenant {
...
+    name,
+  }: {
+    id: string;
+    name: string;
+  }): Promise<EAMTenantByIdResponseDto | null> {
+    const tenant = await this.#tenantRepository.updateById(id, name);
+
+    if (!tenant) {
+      return null;
+    }
```
Comment (Yaroslaww-1) [inline backend/src/services/tenant/tenant.service.ts]:
```suggestion
   const tenant = await repo.getById(id);
   if (!tenant) ...
   tenant.setName(name)
   await repo.save(tenant)
```

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { DEFAULT_PAYLOAD } from '../common/constants';
+import { EAMTenantFormDto, EAMTenantByIdResponseDto } from 'common/types/types';
+
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto | null,
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts:4]:
```suggestion
  tenant: EAMTenantByIdResponseDto,
```

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import { mapTenantDefaultValues } from './helpers/tenant-default-values.helper';
+import styles from './styles.module.scss';
+
+const Tenant: FC = () => {
+  const dispatch = useAppDispatch();
+  const { tenant } = useAppSelector(({ app }) => ({
+    tenant: app.tenant,
+  }));
+
+  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant), [tenant]);
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx:28]:
```suggestion
  const defaultValues = useMemo(() => mapTenantDefaultValues(tenant as EAMTenantByIdResponseDto), [tenant]);
```

File: frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+): EAMTenantFormDto => {
+  const values = {
+    ...DEFAULT_PAYLOAD,
+  };
+  if (tenant) {
+    values.name = tenant.name;
+  }
+  return values;
+};
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts]:
```suggestion
): EAMTenantFormDto => ({
  name: tenant.name,
  // other keys 
});
```

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+} from 'hooks/hooks';
+import { getNameOf } from 'helpers/helpers';
+import {
+  EAMTenantUpdateRequestDto,
+  EAMTenantFormDto,
+  EAMTenantByIdResponseDto,
+} from 'common/types/types';
+import { eamTenantUpdate as EAMTenantValidationSchema } from 'validation-schemas/validation-schemas';
+import { eam as EAMTenantConfigurateActions } from 'store/actions';
+import { mapTenantDefaultValues } from './helpers/tenant-default-values.helper';
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx]:
```suggestion
import { mapTenantDefaultValues } from './helpers/helpers';
```

File: frontend/src/store/eam/common.ts [hunk 18]
```diff
@@ -1,6 +1,7 @@
 enum ActionType {
   GET_WORKERS = 'eam/get-workers',
   LOAD_GROUPS = 'eam/load-groups',
+  UPDATE_TENANT_NAME = 'eam-configurate-tenant/update-tenant-name',
```
Comment (what1s1ove) [inline frontend/src/store/eam/common.ts]:
```suggestion
  UPDATE_TENANT_NAME = 'eam/update-tenant-name',
```

File: frontend/src/store/eam/reducer.ts [hunk 19]
```diff
@@ -4,20 +4,23 @@ import { getWorkers, loadGroups } from './actions';
...
   EAMWorkerGetAllItemResponseDto,
+  EAMTenantByIdResponseDto,
 } from 'common/types/types';
 
 type State = {
   dataStatus: DataStatus;
   workers: EAMWorkerGetAllItemResponseDto[];
   groupsDataStatus: DataStatus;
   groups: EAMGroupGetByTenantResponseItemDto[];
+  tenant: EAMTenantByIdResponseDto | null;
```
Comment (what1s1ove) [inline frontend/src/store/eam/reducer.ts]:
```suggestion
```
we don't want to keep a tenant here. we should keep a tenant in the `app` reducer, but action should for changing should be in the `eam` scoop

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts (hunk 16 above)
Comment (ania-leichenko) [reply frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts]: what if tennant = null?

```jsx
): EAMTenantFormDto => ({
  name: tenant?.name ?? '',
  // other keys 
});
```

File: frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts [hunk 20]
```diff
@@ -0,0 +1,9 @@
+import { EAMTenantFormDto, EAMTenantByIdResponseDto } from 'common/types/types';
+
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto,
+): EAMTenantFormDto => ({
+  name: tenant ? tenant.name : '',
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts]:
```suggestion
  name: tenant?.name ?? '',
```

File: frontend/src/components/eam/components/tenant/tenant.tsx [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+import styles from './styles.module.scss';
+
+const Tenant: FC = () => {
+  const dispatch = useAppDispatch();
+  const { tenant } = useAppSelector(({ app }) => ({
+    tenant: app.tenant,
+  }));
+
+  const defaultValues = useMemo(
+    () => mapTenantDefaultValues(tenant as EAMTenantByIdResponseDto),
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/tenant.tsx]:
```suggestion
    () => mapTenantDefaultValues(tenant),
```
we discussed that we should cast a type

File: frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import { EAMTenantFormDto, EAMTenantByIdResponseDto } from 'common/types/types';
+
+const mapTenantDefaultValues = (
+  tenant: EAMTenantByIdResponseDto,
```
Comment (what1s1ove) [inline frontend/src/components/eam/components/tenant/helpers/tenant-default-values.helper.ts]:
```suggestion
  tenant: EAMTenantByIdResponseDto | null,
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — DISMISSED** [review]

**Review by what1s1ove — DISMISSED** [review]

File: backend/src/services/tenant/tenant.service.ts [hunk 23]
```diff
@@ -32,6 +34,33 @@ class Tenant {
...
+    const tenant = await this.#tenantRepository.getById(id);
+    if (!tenant) {
+      return null;
+    }
+    if (tenant.name === name) {
+      throw new TenantError({
+        status: HttpCode.DENIED,
+        message: ExceptionMessage.TENANT_NAME_EXISTS,
+      });
+    }
```
Comment (Yaroslaww-1) [inline backend/src/services/tenant/tenant.service.ts]: It won't work. Consider an example where we have two tenants:
tenant1 - id1
tenant2 - id2
PUT /id1 with name tenant2
```suggestion
    const tenantWithName = await repo.getByName(name);
    if (tenantWithName !== null) {
      throw new TenantError({
        status: HttpCode.DENIED,
        message: ExceptionMessage.TENANT_NAME_EXISTS,
      });
    }
```

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: backend/src/services/tenant/tenant.service.ts [hunk 24]
```diff
@@ -32,6 +34,34 @@ class Tenant {
...
+  public async updateTenantById({
+    id,
+    name,
+  }: {
+    id: string;
+    name: string;
+  }): Promise<EAMTenantByIdResponseDto | null> {
+    const tenant = await this.#tenantRepository.getById(id);
+    if (!tenant) {
+      return null;
```
Comment (what1s1ove) [inline backend/src/services/tenant/tenant.service.ts]: throw an error here

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by Yaroslaww-1 — DISMISSED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**